tests/Doc/ReferenceTest.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+namespace Castor\Tests\Doc;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\Function_;
+use PhpParser\NodeTraverser;
+use PhpParser\NodeVisitorAbstract;
+use PhpParser\ParserFactory;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Finder\Finder;
+
+class ReferenceTest extends TestCase
+{
+    public function testFunctionsAreReferenced(): void
+    {
+        $functions = $this->getBuiltInFunctions();
+
+        if (!$functions) {
+            $this->fail('Could not find any built-in functions');
+        }
+
+        $lists = $this->getReferencedLists();
+
+        $this->assertNotEmpty($lists);
+        $this->assertArrayHasKey('Functions', $lists);
+        $this->assertIsArray($lists['Functions']);
+        $this->assertNotEmpty($lists['Functions']);
+
+        $referencedFunctions = $lists['Functions'];
+
+        // Ensure all functions are either referenced, internal or deprecated
+        foreach ($functions as $function) {
+            $reflection = new \ReflectionFunction("\\Castor\\{$function}");
+            $comment = $reflection->getDocComment();
+
+            if (str_contains($comment, '@internal')) {
+                $this->fail("Built-in function \"{$function}\" is marked as internal, it should moved to src/functions-internal.php");
+            }
+
+            if (str_contains($comment, '@deprecated')) {
+                $this->assertNotContains($function, $lists['Functions'], "Built-in function \"{$function}\" is deprecated and SHOULD NOT be listed in the documentation doc/reference.md");
+            } else {
+                $this->assertContains($function, $lists['Functions'], "Built-in function \"{$function}\" should be listed in the documentation doc/reference.md");
+            }
+
+            unset($referencedFunctions[array_search($function, $referencedFunctions)]);
+        }
+
+        // Ensure all referenced functions exist
+        foreach ($referencedFunctions as $function) {
+            $this->fail("Built-in function \"{$function}\" is listed in the documentation doc/reference.md but does not exist");
+        }
```

Comment:
I would check the array is empty instead. 
Because if the array is not empty, it'l fail on the the very first loop